### PR TITLE
[249] Crunchbase v3 --> v4 API change and speedups

### LINKED ISSUE #199: Update CB API from v3.1 to v4.0
CSV version 3.1 is set to deprecate on Friday, November 1, 2019 and will no longer be supported. Now is the time to transition to version 4.0 as there will be breaking changes compared to version 3.1. 

See changes in fields:

http://about.crunchbase.com/wp-content/uploads/2019/05/CSV-Update-Changes.pdf

### LINKED ISSUE #249: [HealthMosaic] Crunchbase pipelines break due to new data specification
`OrgCollectTask` fails with

`KeyError: 'category_name'`

from 

```
  File "/home/ec2-user/nesta/nesta/packages/crunchbase/crunchbase_collect.py", line 148, in process_orgs
    cat_groups = cat_groups.set_index(['category_name'])
```

and many warnings along the lines of 

```
WARNING:root:{'id': 'b5907394-9837-f190-7eca-a27462666650', 'name': 'Space Travel', 'type': 'category', 'permalink': 'space-travel-b590', 'cb_url': 'https://www.crunchbase.com/category/space-travel-b590', 'rank': nan, 'created_at': '2014-04-20 07:58:46', 'updated_at': '2018-02-13 00:59:11', 'category_groups_list': 'Transportation'} does not contain any of ['crunchbase_category_groups.category_name'][False]
```

This will probably be quite a lot of work to debug

### PR DESCRIPTION
**Pull request by jaklinger**
Closes #199 

Run

`luigi --module crunchbase_root_task RootTask`

from `nesta/nesta/core/routines/datasets/crunchbase`.

Changes:
- [x] Variable naming changes between v3 and v4
- [x] Debug new warning (described in #249)
- [x] Iterate on naming of non-org collection task
- [x] Change & migration of SQL ORMs
- [x] propagate new fields (see below) to ES

Summary of field changes:

Organisations
--------------

old | new | Change type | New type
-- | -- | -- | --
aliases | alias1 | new field | VARCHAR(200)
  | alias2 | new field | VARCHAR(200)
  | alias3 | new field | VARCHAR(200)
funding_rounds | num_funding_rounds | rename |  
funding_total_us | total_funding_usd | rename |  
  | legal_name | new field | VARCHAR(200)
  | total_funding | new field | BIGINT
  | total_funding_currency_code | new field | VARCHAR(3)
  | num_exits | new field | INTEGER
  | postal_code | new field | VARCHAR(20)
  | rank | new field | INTEGER

Degrees
--------------

old | new | Change type | New type
-- | -- | -- | --
  | cb_url | new field | VARCHAR(200)
degree_id | id | rename |  
  | institution_name | new field | VARCHAR(200)
  | name | new field | VARCHAR(100)
  | permalink | new field | VARCHAR(100)
  | person_name | new field | VARCHAR(100)
  | rank | new field | BIGINT
  | type | new field | VARCHAR(50)

Acquistions
--------------

old | new | Change type | New type
-- | -- | -- | --
state_code | acquiree_state_code | rename |  
  | cb_url | new field | VARCHAR(200)
acquisition_id | id | rename |  
  | name | new field | VARCHAR(100)
  | permalink | new field | VARCHAR(100)
  | rank | new field | BIGINT
  | type | new field | VARCHAR(50)

Funding Rounds
------------------

old | new | Change type
-- | -- | --
funding_round_id | id | rename
  | name | new field
investor_ids | lead_investor_ids | rename
investor_names |   | removed
post_money_currency_code | post_money_valuation_currency_code | rename
  | type | new field
  | rank | new field
company_id | org_id | rename
company_name | org_name | rename
  | permalink | new field

Funds
------------------

old | new | Change type
-- | -- | --
fund_id | id | rename
fund_name | name | rename
  | raised_amount_usd | new field
  | cb_url | new field
  | entity_name | new field
  | entity_type | new field
  | permalink | new field
  | rank | new field
  | type | new field

Investment partners
-----------------------

old | new | Change type
-- | -- | --
  | funding_round_name | new field
  | investor_name | new field
  | name | new field
  | id | new field
  | partner_name | new field
  | permalink | new field
  | rank | new field
  | type | new field
  | updated_at | new field
  | cb_url | new field
  | created_at | new field

Investments
------------------

old | new | Change type
-- | -- | --
  | funding_round_name | new field
  | id | new field
  | investor_name | new field
  | investor_type | new field
  | name | new field
  | permalink | new field
  | rank | new field
  | type | new field
  | updated_at | new field
  | cb_url | new field
  | created_at | new field

Investors
------------------

old | new | Change type
-- | -- | --
country | country_code | rename
investor_name | name | rename
investor_type | investor_types | rename
  | total_funding | new field
  | total_funding_currency_code | new field
  | created_at | new field
  | linkedin_url | new field
  | permalink | new field
  | rank | new field
  | type | new field

IPOs
------------------

old | new | Change type
-- | -- | --
company_id | org_id | rename
company_state_code | state_code | rename
country | country_code | rename
ipo_id | id | rename
price | share_price | rename
price_currency_code | share_price_currency_code | rename
price_usd | share_price_usd | rename
  | type | rename
  | valuation_price | new field
  | valuation_price_currency_code | new field
  | valuation_price_usd | new field
  | money_raised | new field
  | money_raised_currency_code | new field
  | org_cb_url | new field
  | org_name | new field
  | permalink | new field
  | rank | new field

People
------------------

old | new | Change type
-- | -- | --
job_id | id | rename
  | name | new field
  | org_name | new field
  | permalink | new field
  | person_name | new field
  | rank | new field
  | type | new field
  | updated_at | new field
  | cb_url | new field
  | created_at | new field

Jobs
------------------

old | new | Change type
-- | -- | --
primary_affiliation_organization | featured_job_organization_name | rename
primary_affiliation_title | featured_job_title | rename
primary_organization_id | featured_job_organization_id | rename
  | name | new field
  | permalink | new field
  | rank | new field
  | region | new field
  | type | new field

### CODE REVIEW

File: nesta/core/luigihacks/batchgeocode.py [hunk 1]
```diff
@@ -35,10 +36,10 @@ class GeocodeBatchTask(AutoBatchTask):
     test = luigi.BoolParameter()
     _routine_id = luigi.Parameter(default="DUMMY ROUTINE")
     db_config_env = luigi.Parameter()
-    city_col = luigi.Parameter()
-    country_col = luigi.Parameter()
+    city_col = SqlAlchemyParameter()
```
Comment (jaklinger) [inline nesta/core/luigihacks/batchgeocode.py:39]: gets rid of annoying warnings since it is assumed that `luigi.Parameter` is  `str`

File: nesta/core/routines/datasets/crunchbase/crunchbase_geocode_task.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+'''
```
Comment (jaklinger) [inline nesta/core/routines/datasets/crunchbase/crunchbase_geocode_task.py:1]: This file was put in since Russ's implementation ran all 5 geocoding tasks sequentially, whereas now they run in parallel (see crunchbase_root_task.py)

File: nesta/core/routines/projects/health_mosaic/crunchbase_elasticsearch_task.py [hunk 3]
```diff
@@ -14,7 +14,7 @@
 import luigi
```
Comment (jaklinger) [inline nesta/core/routines/projects/health_mosaic/crunchbase_elasticsearch_task.py:14]: Note for this, and all `health_mosaic` files: I've factored out core crunchbase collection from the health-mosaic pipeline to make the routine more general purpose. the `health_mosaic` codebase has been unmaintained for several months anyway, so this will get picked up in future PRs anyway, if applicable

File: nesta/packages/crunchbase/crunchbase_collect.py [hunk 4]
```diff
@@ -69,8 +69,11 @@ def get_files_from_tar(files, nrows=None):
     dfs = []
     with crunchbase_tar() as tar:
         for filename in files:
```
Comment (jaklinger) [inline nesta/packages/crunchbase/crunchbase_collect.py:71]: previously all dealings with `null` values from the CSV were dealt with in individual functions. I've refactored so that a consistent procedure is performed for all crunchbase CSVs

File: nesta/packages/crunchbase/crunchbase_collect.py [hunk 5]
```diff
@@ -135,17 +138,12 @@ def process_orgs(orgs, existing_orgs, cat_groups, org_descriptions):
...
-
     # lookup country name and add as a column
     orgs['country'] = orgs['country_code'].apply(country_iso_code_to_name)
-    orgs = orgs.drop('country_code', axis=1)  # now redundant with country_alpha_3 appended
-
+    
     orgs['long_description'] = None
     org_cats = []
-    cat_groups = cat_groups.set_index(['category_name'])
+    cat_groups = cat_groups.set_index(['name'])
```
Comment (jaklinger) [inline nesta/packages/crunchbase/crunchbase_collect.py:146]: example of variable renaming between versions 3 and 4

File: nesta/core/orms/crunchbase_orm.py [hunk 6]
```diff
@@ -11,241 +11,357 @@
 
 Base = declarative_base()
 
+# There are a lot of repeated data types in this schema, so the following
+# fixtures are designed to maintain consistency between tables
+_TEXT = TEXT(collation='utf8mb4_unicode_ci')
```
Comment (bishax) [inline nesta/core/orms/crunchbase_orm.py:16]: Out of curiosity rather than review neccessity: Why the change to `utf8mb4_unicode_ci`? Avoid MySQL 3 byte UTF-8 encoding? (c.f. https://www.eversql.com/mysql-utf8-vs-utf8mb4-whats-the-difference-between-utf8-and-utf8mb4/)

File: nesta/core/orms/crunchbase_orm.py [hunk 7]
```diff
@@ -11,241 +11,357 @@
 
 Base = declarative_base()
 
+# There are a lot of repeated data types in this schema, so the following
+# fixtures are designed to maintain consistency between tables
```
Comment (bishax) [inline nesta/core/orms/crunchbase_orm.py:15]: Even without repetition it makes reading a lot nicer :+1:

**Review by bishax — APPROVED** [review]
`This PR looks :)`

File: nesta/core/orms/crunchbase_orm.py (hunk 6 above)
Comment (jaklinger) [reply nesta/core/orms/crunchbase_orm.py:16]: the main use case is emoticons, which crop up from time-to-time, and also some reasonably looking non-ascii characters appear in the data. Crazy that it's not default encoding!
